test/test_nn.py

```diff
@@ -59,6 +59,20 @@ def wrapper(*args, **kwargs):
     return decorator
 
 
+def _assertGradAndGradgradChecks(assertTrueFn, apply_fn, inputs):
```

Comment:
nit: it might be nicer to pass the test case object, not its method. It's harder to make an error this way (you only have to choose an appropriate method within `_assertGradAndGradgradChecks` not in every place where you call it).